src/core/SpyRegistry.js

```diff
@@ -175,18 +175,41 @@ getJasmineRequireObj().SpyRegistry = function(j$) {
         prop,
         descriptor;
 
-      while (pointer) {
-        for (prop in pointer) {
-          if (
-            Object.prototype.hasOwnProperty.call(pointer, prop) &&
-            pointer[prop] instanceof Function
+      while (
+        pointer &&
+        (!includeNonEnumerable || pointer !== Object.prototype)
+      ) {
+        var enumerableProperties = Object.keys(pointer);
+        var properties;
+        if (includeNonEnumerable) {
+          properties = Object.getOwnPropertyNames(pointer).filter(function(
+            prop
           ) {
-            descriptor = Object.getOwnPropertyDescriptor(pointer, prop);
-            if (
-              (descriptor.writable || descriptor.set) &&
-              descriptor.configurable
-            ) {
-              props.push(prop);
+            return (
+              prop !== 'constructor' ||
+              enumerableProperties.indexOf('constructor') > -1
+            );
+          });
+        } else {
+          properties = enumerableProperties;
+        }
+        for (var i = 0; i < properties.length; i++) {
+          prop = properties[i];
+          if (Object.prototype.hasOwnProperty.call(pointer, prop)) {
+            var value;
+            try {
+              value = pointer[prop];
+            } catch (e) {
+              /* tried to access a property that should not be accessed */
```

Comment:
I missed it the first time through, but there's a subtle bug here. Or possibly several subtle bugs interacting with each other. When I run the tests on Safari 13 and 14:

```
Failures:
1) SpyRegistry #spyOnAllFunctions when includeNonEnumerable is true should not throw an exception if we try and access strict mode restricted properties
  Message:
    Expected function not to throw, but it threw Error: <spyOn> : caller is not declared writable or has no setter
    Usage: spyOn(<object>, <methodName>).
  Stack:
    <Jasmine>
    http://localhost:53530/__spec__/core/SpyRegistrySpec.js:498:31
    <Jasmine>
```

I think there's at least two things going on. The main issue is that certain properties of functions on Safari are defined as completely inaccessible on the function itself but as read-only higher up the prototype chain. The right thing to do is to treat the first descriptor found for a given property as definitive and ignore any others that are further up the prototype chain, but this code gives a read-write property descriptor on an object's prototype precedence over a read-write property descriptor on the object itself. If you don't have Safari handy you should be able to reproduce the problem by setting up a similar object graph using `Object.create` and `Object.defineProperty`, and that could be the foundation for a unit test that exercises this case in all browsers.

The other issue is that if we catch an exception on line 202, we continue on with the rest of the loop reusing the value of the `value` variable from the previous iteration. I don't know if that's currently causing any visible breakage but it's definitely not right. There's three things I'd like to see here:

1. Break out of the loop from the catch block.
2. Consider extracting parts of this into smaller functions, if that will make the flow control easier to understand.
3. Move all `var` statements to the beginning of the function, to match their actual scope and avoid the suggestion that they have block scope like variables in other languages.